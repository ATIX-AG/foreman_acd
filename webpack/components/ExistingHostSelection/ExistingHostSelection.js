import React from 'react';
import PropTypes from 'prop-types';
import { cloneDeep } from 'lodash';
import { DualListControlled } from 'patternfly-react';
import { translate as __ } from 'foremanReact/common/I18n';
import ServiceSelector from './components/ServiceSelector';
import { arrayToObject } from '../../helper';

class ExistingHostSelection extends React.Component {
  componentDidMount() {
    const { allHosts, initExistingHostSelection } = this.props;

    initExistingHostSelection(allHosts);
  }

  render() {
    const {
      services,
      serviceId,
      availableHosts,
      alreadyUsedHosts,
      loadHostsOfHostgroup,
      hostSelectionChanged,
    } = this.props;

    const serviceList = arrayToObject(services, 'id', 'name');
    const loadHostgroupUrl = '/api/v2/hostgroups/__hostgroup_id__/hosts';

    return (
      <div>
        <div className="row">
          <ServiceSelector
            label="Service"
            hidden={false}
            selectValue={serviceId ? serviceId.toString() : '0'}
            options={serviceList}
            onChange={loadHostsOfHostgroup}
            additionalData={{ url: loadHostgroupUrl, services }}
          />
        </div>
        <div className="row">
          <label className="col-md-2 control-label">{__('Hosts')}</label>
          {serviceId !== undefined ? (
            <div className="col-md-6">
              <DualListControlled
                onChange={hostSelectionChanged}
                left={{
                  items: cloneDeep(availableHosts),
                }}
                right={{
                  items: cloneDeep(alreadyUsedHosts),
                }}
                allowHiddenInputs={false}
              />
            </div>
          ) : (
            <span>{__('Please select service first.')}</span>
          )}
        </div>
      </div>
    );
  }
}

ExistingHostSelection.defaultProps = {
  allHosts: [],
  serviceId: undefined,
  availableHosts: [],
  alreadyUsedHosts: [],
};

ExistingHostSelection.propTypes = {
  allHosts: PropTypes.array,
  services: PropTypes.array.isRequired,
  initExistingHostSelection: PropTypes.func.isRequired,
  serviceId: PropTypes.number,
  availableHosts: PropTypes.array,
  alreadyUsedHosts: PropTypes.array,
  loadHostsOfHostgroup: PropTypes.func.isRequired,
  hostSelectionChanged: PropTypes.func.isRequired,
};

export default ExistingHostSelection;
