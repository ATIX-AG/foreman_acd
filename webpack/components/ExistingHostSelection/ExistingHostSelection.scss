@import '~@theforeman/vendor/scss/variables';

.add_existing_hosts_modal {
    margin-top: 0;
    margin-bottom: 0;
    height: 100%;
    width: 80%;
    max-width: none !important;
}

.row {
  margin-left: 0;
  margin-right: 0;
  margin-bottom: 20px;
}
