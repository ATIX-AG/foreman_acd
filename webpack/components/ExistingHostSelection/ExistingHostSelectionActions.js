import React from 'react';
import * as sort from 'sortabular';
import { translate as __ } from 'foremanReact/common/I18n';

import api from 'foremanReact/API';

import {
  propsToSnakeCase,
  propsToCamelCase,
} from 'foremanReact/common/helpers';

import {
  EXISTING_HOST_SELECTION_INIT,
  EXISTING_HOST_SELECTION_LOAD_HOSTS_SUCCESS,
  EXISTING_HOST_SELECTION_LOAD_HOSTS_FAILURE,
  EXISTING_HOST_SELECTION_SELECTION_CHANGED,
} from './ExistingHostSelectionConstants';

export const initExistingHostSelection = allHosts => dispatch => {
  const initialState = {};

  initialState.alreadyUsedHosts = undefined;
  initialState.serviceId = undefined;
  initialState.selectedHosts = [];
  initialState.hostsInHostgroup = {};
  initialState.allHosts = allHosts;

  dispatch({
    type: EXISTING_HOST_SELECTION_INIT,
    payload: initialState,
  });
};

export const loadHostsOfHostgroup = (serviceId, additionalData) => dispatch => {
  const selService = additionalData.services.filter(s => s.id == serviceId)[0];
  const realUrl = additionalData.url.replace(
    '__hostgroup_id__',
    selService.hostgroup
  );

  return api
    .get(realUrl, {}, {})
    .then(({ data }) =>
      dispatch({
        type: EXISTING_HOST_SELECTION_LOAD_HOSTS_SUCCESS,
        payload: {
          hosts: data.results,
          serviceId: parseInt(serviceId),
        },
      })
    )
    .catch(error =>
      dispatch(errorHandler(EXISTING_HOST_SELECTION_LOAD_HOSTS_FAILURE, error))
    );
};

export const hostSelectionChanged = ({ left, right }) => ({
  type: EXISTING_HOST_SELECTION_SELECTION_CHANGED,
  payload: {
    selection: right.items,
  },
});

const errorHandler = (msg, err) => {
  const error = {
    errorMsg: __('Failed to fetch data from server.'),
    statusText: err,
  };
  return { type: msg, payload: { error } };
};
