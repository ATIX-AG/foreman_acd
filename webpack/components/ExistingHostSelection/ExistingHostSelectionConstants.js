export const EXISTING_HOST_SELECTION_INIT = 'EXISTING_HOST_SELECTION_INIT';
export const EXISTING_HOST_SELECTION_LOAD_HOSTS_SUCCESS =
  'EXISTING_HOST_SELECTION_LOAD_HOSTS_SUCCESS';
export const EXISTING_HOST_SELECTION_LOAD_HOSTS_FAILURE =
  'EXISTING_HOST_SELECTION_LOAD_HOSTS_FAILURE';
export const EXISTING_HOST_SELECTION_SELECTION_CHANGED =
  'EXISTING_HOST_SELECTION_SELECTION_CHANGED';
