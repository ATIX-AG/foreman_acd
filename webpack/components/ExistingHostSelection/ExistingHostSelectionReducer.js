import Immutable from 'seamless-immutable';

import {
  cloneDeep,
  findIndex,
  findLastIndex,
} from 'lodash';

import { shortHostname } from '../../helper';
import * as sort from 'sortabular';

import {
  EXISTING_HOST_SELECTION_INIT,
  EXISTING_HOST_SELECTION_LOAD_HOSTS_SUCCESS,
  EXISTING_HOST_SELECTION_LOAD_HOSTS_FAILURE,
  EXISTING_HOST_SELECTION_SELECTION_CHANGED,
} from './ExistingHostSelectionConstants';

export const initialState = Immutable({
  serviceId: undefined,
  error: { errorMsg: '', status: '', statusText: '' },
});

const existingHostSelectionConf = (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case EXISTING_HOST_SELECTION_INIT: {
      return state.merge(payload);
    }
    case EXISTING_HOST_SELECTION_LOAD_HOSTS_SUCCESS: {
      let alreadyUsedHosts = [];

      console.log(state.allHosts);
      console.log(payload);

      state.allHosts.forEach(host => {
        if ((host.service == payload.serviceId) && (host.isExistingHost == true)) {
          alreadyUsedHosts.push({
            value: host.hostname,
            label: host.hostname,
            disabled: true,
            tooltipText: __('Host already used for this application instance'),
          });
        }
      });

      let availableHosts = [];
      let hostsInHostgroup = {};
      payload.hosts.forEach(host => {
        let shortName = shortHostname(host.name);
        if (state.allHosts.find(h => h.hostname == shortName) == undefined) {
          availableHosts.push({
            value: shortName,
            label: shortName,
          });
        }
        hostsInHostgroup[shortName] = host;
      });

      return state.merge({
        hostsInHostgroup: hostsInHostgroup,
        availableHosts: availableHosts,
        alreadyUsedHosts: alreadyUsedHosts,
        serviceId: payload.serviceId,
      });
    }
    case EXISTING_HOST_SELECTION_LOAD_HOSTS_FAILURE: {
      return state.merge({ error: payload.error });
    }
    case EXISTING_HOST_SELECTION_SELECTION_CHANGED: {
      let selectedHosts = [];

      payload.selection.forEach(selHost => {
        let hostData = state.hostsInHostgroup[selHost.value];
        selectedHosts.push({
          fqdn: hostData.name,
          hostname: shortHostname(hostData.name),
          hostgroupId: hostData.hostgroup_id,
          serviceId: state.serviceId,
        });
      });

      return state.merge({
        selectedHosts: selectedHosts,
      });
    }
    default:
      return state;
  }
};

export default existingHostSelectionConf;
