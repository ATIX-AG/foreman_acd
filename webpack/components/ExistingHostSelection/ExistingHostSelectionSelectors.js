const existingHostState = state => state.foremanAcd.existingHostSelectionConf;

export const selectServiceId = state => existingHostState(state).serviceId;
export const selectAllHosts = state => existingHostState(state).allHosts;
export const selectHostsInHostgroup = state => existingHostState(state).selectHostsInHostgroup;
export const selectAvailableHosts = state => existingHostState(state).availableHosts;
export const selectAlreadyUsedHosts = state => existingHostState(state).alreadyUsedHosts;
export const selectSelectedHosts = state => existingHostState(state).selectedHosts;
