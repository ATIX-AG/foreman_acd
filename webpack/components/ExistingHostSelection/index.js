import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import './ExistingHostSelection.scss';
import ExistingHostSelection from './ExistingHostSelection';
import * as ExistingHostSelectionActions from './ExistingHostSelectionActions';

import {
  selectServiceId,
  selectAvailableHosts,
  selectAlreadyUsedHosts,
} from './ExistingHostSelectionSelectors';

const mapStateToProps = state => ({
  serviceId: selectServiceId(state),
  availableHosts: selectAvailableHosts(state),
  alreadyUsedHosts: selectAlreadyUsedHosts(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(ExistingHostSelectionActions, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExistingHostSelection);
