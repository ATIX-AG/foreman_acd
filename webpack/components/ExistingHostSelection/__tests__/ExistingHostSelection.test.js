import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import ExistingHostSelection from '../ExistingHostSelection';

const noop = () => {};

const fixtures = {
  'should render existing host selection': {
    location: 'Default Location',
    organization: 'Default Organization',
    services: [],
    allHosts: [],
    initExistingHostSelection: noop,
    loadHostsOfHostgroup: noop,
  },
};

describe('ExistingHostSelection', () =>
  testComponentSnapshotsWithFixtures(ExistingHostSelection, fixtures));
