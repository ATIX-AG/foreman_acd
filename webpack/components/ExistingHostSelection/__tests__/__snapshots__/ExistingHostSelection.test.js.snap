// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExistingHostSelection should render existing host selection 1`] = `
<div>
  <div
    className="row"
  >
    <ServiceSelector
      additionalData={
        Object {
          "services": Array [],
          "url": "/api/v2/hostgroups/__hostgroup_id__/hosts",
        }
      }
      hidden={false}
      label="Service"
      onChange={[Function]}
      options={Object {}}
      selectValue="0"
    />
  </div>
  <div
    className="row"
  >
    <label
      className="col-md-2 control-label"
    >
      Hosts
    </label>
    <span>
      Please select service first.
    </span>
  </div>
</div>
`;
